File: Rust/src/_1_two_sum.rs
use std::time::{Duration, Instant};

pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut map = std::collections::HashMap::new();
    for (i, num) in nums.iter().enumerate() {
        let complement = target - num;
        if let Some(&j) = map.get(&complement) {
            return vec![j as i32, i as i32];
        }
        map.insert(num, i);
    }
    vec![]
}

pub(crate)fn test_two_sum(nums: Vec<i32>, target: i32) {
    println!("nums = {:?}" , nums);
    println!("target = {:?}" , target);

    let start: Instant = Instant::now();
    let result: Vec<i32> = two_sum(nums, target);
    let duration: Duration = start.elapsed();

    println!("result = {:?}", result);
    println!("duration = {:?}", duration);
}

File: Go/4_arrays/1_Two_Sum.go
package main

import (
	"fmt"
	"time"
)

func twoSum(nums []int, target int) []int {
	m := make(map[int]int)

	for i, num := range nums {
		complement := target - num
		if j, ok := m[complement]; ok {
			return []int{j, i}
		}
		m[num] = i
	}
	return nil
}

func testTwoSum(nums []int, target int) {
	fmt.Println("List:	 ", nums)
	fmt.Println("Target: ", target)

	start := time.Now()
	result := twoSum(nums, target)
	elapsed := time.Since(start)

	fmt.Println("Result: ", result)
	fmt.Println("Execution time: ", elapsed.Nanoseconds())
}

func main() {
	list1 := []int{2, 7, 11, 15}
	target1 := 9
	testTwoSum(list1, target1)

	list2 := []int{3, 2, 4}
	target2 := 6
	testTwoSum(list2, target2)
}
